File: android/src/main/java/expo/modules/callkittelecom/ExpoCallKitTelecomModule.kt
package expo.modules.callkittelecom

import expo.modules.callkittelecom.events.CallEventEmitter
import expo.modules.callkittelecom.events.CallEvents
import expo.modules.callkittelecom.managers.CallAudioManager
import expo.modules.callkittelecom.managers.CallManager
import expo.modules.callkittelecom.managers.CaptureSessionManager
import expo.modules.callkittelecom.managers.VoIPPushManager
import expo.modules.callkittelecom.models.CallEndedReason
import expo.modules.callkittelecom.models.CallOptions
import expo.modules.callkittelecom.models.CallParticipant
import expo.modules.callkittelecom.models.IncomingCallEvent
import expo.modules.callkittelecom.services.CallNotificationReceiver
import expo.modules.callkittelecom.store.CallStore
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

class ExpoCallKitTelecomModule : Module() {
    /**
     * Checks the current activity's launch intent for ACTION_ANSWER.
     *
     * On cold start from a notification answer action, the intent is delivered to onCreate() — not
     * onNewIntent() — so OnNewIntent never fires. This method handles that case by checking the
     * launch intent directly.
     */
    private fun handleLaunchIntent() {
        val intent = appContext.currentActivity?.intent ?: return
        handleAnswerIntent(intent)
    }

    /** Extracts the call ID from an ACTION_ANSWER intent and answers the call. */
    private fun handleAnswerIntent(intent: android.content.Intent) {
        if (intent.action != CallNotificationReceiver.ACTION_ANSWER) return

        val callIdStr = intent.getStringExtra(CallNotificationReceiver.EXTRA_CALL_ID) ?: return
        val callId =
            try {
                UUID.fromString(callIdStr)
            } catch (_: IllegalArgumentException) {
                return
            }

        // Clear the action so it isn't re-processed on configuration changes
        intent.action = null

        CallManager.shared.answerCall(callId)
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoCallKitTelecom")

        // region Events

        Events(
            CallEvents.CALL_SESSION_ADDED,
            CallEvents.CALL_SESSION_UPDATED,
            CallEvents.CALL_SESSION_REMOVED,
            CallEvents.AUDIO_SESSION_ACTIVATED,
            CallEvents.AUDIO_SESSION_DEACTIVATED,
            CallEvents.AUDIO_ROUTE_CHANGED,
            CallEvents.INCOMING_CALL_REPORTED,
            CallEvents.OUTGOING_CALL_STARTED,
            CallEvents.CALL_ANSWERED,
            CallEvents.CALL_ENDED,
            CallEvents.CALL_REPORTED_ENDED,
            CallEvents.SET_MUTED_ACTION,
            CallEvents.VIDEO_CHANGED,
            CallEvents.SET_HELD_ACTION,
            CallEvents.DTMF,
            CallEvents.CALL_INTENT_RECEIVED,
            CallEvents.VOIP_PUSH_TOKEN_UPDATED,
        )

        // endregion

        // region Lifecycle

        OnCreate {
            val context = appContext.reactContext ?: return@OnCreate

            CallManager.shared.initialize(context)

            VoIPPushManager.register()

            CallEventEmitter.setSender { eventName, body -> sendEvent(eventName, body) }

            // Handle the launch intent for cold-start answer actions.
            // OnNewIntent only fires when the activity is already running.
            // On cold start the ACTION_ANSWER intent is the launch intent,
            // so we must check it here.
            handleLaunchIntent()
        }

        OnDestroy { CallEventEmitter.setSender(null) }

        OnNewIntent { intent -> handleAnswerIntent(intent) }

        // Register per-event observers
        OnStartObserving(CallEvents.CALL_SESSION_ADDED) {
            CallEventEmitter.startObserving(CallEvents.CALL_SESSION_ADDED)
        }
        OnStopObserving(CallEvents.CALL_SESSION_ADDED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_SESSION_ADDED)
        }

        OnStartObserving(CallEvents.CALL_SESSION_UPDATED) {
            CallEventEmitter.startObserving(CallEvents.CALL_SESSION_UPDATED)
        }
        OnStopObserving(CallEvents.CALL_SESSION_UPDATED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_SESSION_UPDATED)
        }

        OnStartObserving(CallEvents.CALL_SESSION_REMOVED) {
            CallEventEmitter.startObserving(CallEvents.CALL_SESSION_REMOVED)
        }
        OnStopObserving(CallEvents.CALL_SESSION_REMOVED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_SESSION_REMOVED)
        }

        OnStartObserving(CallEvents.AUDIO_SESSION_ACTIVATED) {
            CallEventEmitter.startObserving(CallEvents.AUDIO_SESSION_ACTIVATED)
        }
        OnStopObserving(CallEvents.AUDIO_SESSION_ACTIVATED) {
            CallEventEmitter.stopObserving(CallEvents.AUDIO_SESSION_ACTIVATED)
        }

        OnStartObserving(CallEvents.AUDIO_SESSION_DEACTIVATED) {
            CallEventEmitter.startObserving(CallEvents.AUDIO_SESSION_DEACTIVATED)
        }
        OnStopObserving(CallEvents.AUDIO_SESSION_DEACTIVATED) {
            CallEventEmitter.stopObserving(CallEvents.AUDIO_SESSION_DEACTIVATED)
        }

        OnStartObserving(CallEvents.AUDIO_ROUTE_CHANGED) {
            CallEventEmitter.startObserving(CallEvents.AUDIO_ROUTE_CHANGED)
        }
        OnStopObserving(CallEvents.AUDIO_ROUTE_CHANGED) {
            CallEventEmitter.stopObserving(CallEvents.AUDIO_ROUTE_CHANGED)
        }

        OnStartObserving(CallEvents.INCOMING_CALL_REPORTED) {
            CallEventEmitter.startObserving(CallEvents.INCOMING_CALL_REPORTED)
        }
        OnStopObserving(CallEvents.INCOMING_CALL_REPORTED) {
            CallEventEmitter.stopObserving(CallEvents.INCOMING_CALL_REPORTED)
        }

        OnStartObserving(CallEvents.OUTGOING_CALL_STARTED) {
            CallEventEmitter.startObserving(CallEvents.OUTGOING_CALL_STARTED)
        }
        OnStopObserving(CallEvents.OUTGOING_CALL_STARTED) {
            CallEventEmitter.stopObserving(CallEvents.OUTGOING_CALL_STARTED)
        }

        OnStartObserving(CallEvents.CALL_ANSWERED) {
            CallEventEmitter.startObserving(CallEvents.CALL_ANSWERED)
        }
        OnStopObserving(CallEvents.CALL_ANSWERED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_ANSWERED)
        }

        OnStartObserving(CallEvents.CALL_ENDED) {
            CallEventEmitter.startObserving(CallEvents.CALL_ENDED)
        }
        OnStopObserving(CallEvents.CALL_ENDED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_ENDED)
        }

        OnStartObserving(CallEvents.CALL_REPORTED_ENDED) {
            CallEventEmitter.startObserving(CallEvents.CALL_REPORTED_ENDED)
        }
        OnStopObserving(CallEvents.CALL_REPORTED_ENDED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_REPORTED_ENDED)
        }

        OnStartObserving(CallEvents.SET_MUTED_ACTION) {
            CallEventEmitter.startObserving(CallEvents.SET_MUTED_ACTION)
        }
        OnStopObserving(CallEvents.SET_MUTED_ACTION) {
            CallEventEmitter.stopObserving(CallEvents.SET_MUTED_ACTION)
        }

        OnStartObserving(CallEvents.VIDEO_CHANGED) {
            CallEventEmitter.startObserving(CallEvents.VIDEO_CHANGED)
        }
        OnStopObserving(CallEvents.VIDEO_CHANGED) {
            CallEventEmitter.stopObserving(CallEvents.VIDEO_CHANGED)
        }

        OnStartObserving(CallEvents.SET_HELD_ACTION) {
            CallEventEmitter.startObserving(CallEvents.SET_HELD_ACTION)
        }
        OnStopObserving(CallEvents.SET_HELD_ACTION) {
            CallEventEmitter.stopObserving(CallEvents.SET_HELD_ACTION)
        }

        OnStartObserving(CallEvents.DTMF) { CallEventEmitter.startObserving(CallEvents.DTMF) }
        OnStopObserving(CallEvents.DTMF) { CallEventEmitter.stopObserving(CallEvents.DTMF) }

        OnStartObserving(CallEvents.CALL_INTENT_RECEIVED) {
            CallEventEmitter.startObserving(CallEvents.CALL_INTENT_RECEIVED)
        }
        OnStopObserving(CallEvents.CALL_INTENT_RECEIVED) {
            CallEventEmitter.stopObserving(CallEvents.CALL_INTENT_RECEIVED)
        }

        OnStartObserving(CallEvents.VOIP_PUSH_TOKEN_UPDATED) {
            CallEventEmitter.startObserving(CallEvents.VOIP_PUSH_TOKEN_UPDATED)
        }
        OnStopObserving(CallEvents.VOIP_PUSH_TOKEN_UPDATED) {
            CallEventEmitter.stopObserving(CallEvents.VOIP_PUSH_TOKEN_UPDATED)
        }

        // endregion

        // region Call Session

        // Returns the first active call session, if present.
        AsyncFunction("getActiveCallSession") { CallStore.firstSession()?.toMap() }

        // endregion

        // region Audio Session

        // Returns current audio session snapshot for diagnostics/UI state.
        Function("getAudioSessionState") { CallAudioManager.getAudioSessionState() }

        // Prepares audio session state before starting/reporting a call.
        Function("prepareAudioSessionForCall") { hasVideo: Boolean ->
            CallAudioManager.prepareAudioSessionForCall(hasVideo)
        }

        // Restores pre-call audio mode/route state.
        Function("restoreAudioSession") { CallAudioManager.restoreAudioSession() }

        // Overrides route to speaker (true) or clears override (false).
        Function("setAudioSessionPortOverride") { enabled: Boolean ->
            CallAudioManager.setAudioSessionPortOverride(enabled)
        }

        // endregion

        // region Capture Session

        // Returns capture session state (currently camera permission).
        Function("getCaptureSessionState") { CaptureSessionManager.getCaptureSessionState() }

        // endregion

        // region Start Outgoing Call

        // Starts a new outgoing call and returns the native call session ID.
        AsyncFunction("startOutgoingCall") {
            recipient: Map<String, Any?>,
            options: Map<String, Any?> ->
            val callId =
                CallManager.shared.startOutgoingCall(
                    recipient = CallParticipant.fromMap(recipient),
                    options = CallOptions(hasVideo = options["hasVideo"] as? Boolean ?: false),
                )
            callId
        }

        // endregion

        // region Report Incoming Call

        // Reports an incoming call to Telecom from app/push events.
        AsyncFunction("reportIncomingCall") { event: Map<String, Any?> ->
            CallManager.shared.reportIncomingCall(IncomingCallEvent.fromMap(event))
        }

        // endregion

        // region Answer Call

        // Answers an existing incoming call session.
        AsyncFunction("answerCall") { id: String ->
            CallManager.shared.answerCall(UUID.fromString(id))
        }

        // Fulfills pending incoming-call answer once media is connected.
        AsyncFunction("fulfillIncomingCallAnswered") { requestId: String ->
            CallManager.shared.fulfillIncomingCallConnected(UUID.fromString(requestId))
        }

        // Reports outgoing call media connection established.
        AsyncFunction("reportOutgoingCallConnected") { id: String ->
            CallManager.shared.reportOutgoingCallConnected(UUID.fromString(id))
        }

        // endregion

        // region End Call

        // Ends an active call.
        AsyncFunction("endCall") { id: String -> CallManager.shared.endCall(UUID.fromString(id)) }

        // Reports an externally-ended call with explicit reason.
        AsyncFunction("reportCallEnded") { id: String, reason: String ->
            CallManager.shared.reportCallEnded(
                UUID.fromString(id),
                CallEndedReason.fromValue(reason),
            )
        }

        // endregion

        // region Mute Support

        // Sets mute state for a call.
        AsyncFunction("setMuted") { id: String, muted: Boolean ->
            CallManager.shared.setMuted(UUID.fromString(id), muted)
        }

        // endregion

        // region Video Support

        // Reports call video enabled state changes.
        AsyncFunction("reportVideo") { id: String, enabled: Boolean ->
            CallManager.shared.reportVideo(UUID.fromString(id), enabled)
        }

        // endregion

        // region Hold Support

        // Sets call hold state.
        AsyncFunction("setHeld") { id: String, onHold: Boolean ->
            CallManager.shared.setHeld(UUID.fromString(id), onHold)
        }

        // endregion

        // region DTMF Support

        // Sends requested DTMF digits for a call.
        AsyncFunction("playDTMF") { id: String, digits: String ->
            CallManager.shared.playDTMF(UUID.fromString(id), digits)
        }

        // endregion

        // region VoIP Push

        // Registers for VoIP push by fetching the FCM token.
        Function("registerVoIPPush") { VoIPPushManager.register() }

        // Returns the current VoIP push token and its type.
        Function("getVoIPPushToken") { mapOf("token" to VoIPPushManager.token, "type" to "FCM") }

        // endregion
    }
}

File: ios/ExpoCallKitTelecomModule.swift
import CallKit
import ExpoModulesCore

public class ExpoCallKitTelecomModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoCallKitTelecom")

    // MARK: - Events

    Events(
      CallSessionAddedEvent.name,
      CallSessionUpdatedEvent.name,
      CallSessionRemovedEvent.name,
      AudioSessionActivatedEvent.name,
      AudioSessionDeactivatedEvent.name,
      AudioRouteChangedEvent.name,
      IncomingCallReportedEvent.name,
      OutgoingCallStartedEvent.name,
      CallAnsweredEvent.name,
      CallEndedEvent.name,
      CallReportedEnded.name,
      SetMutedActionEvent.name,
      VideoChangedEvent.name,
      SetHeldActionEvent.name,
      DTMFEvent.name,
      CallIntentReceivedEvent.name,
      VoIPPushTokenUpdatedEvent.name
    )

    // MARK: - Lifecycle

    OnCreate { [weak self] in
      Task { @MainActor in
        CallEventEmitter.shared.module = self
      }
    }

    // Register per-event observers
    OnStartObserving(CallSessionAddedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallSessionAddedEvent.name
        )
      }
    }
    OnStopObserving(CallSessionAddedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallSessionAddedEvent.name
        )
      }
    }

    OnStartObserving(CallSessionUpdatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallSessionUpdatedEvent.name
        )
      }
    }
    OnStopObserving(CallSessionUpdatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallSessionUpdatedEvent.name
        )
      }
    }

    OnStartObserving(CallSessionRemovedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallSessionRemovedEvent.name
        )
      }
    }
    OnStopObserving(CallSessionRemovedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallSessionRemovedEvent.name
        )
      }
    }

    OnStartObserving(AudioSessionActivatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: AudioSessionActivatedEvent.name
        )
      }
    }
    OnStopObserving(AudioSessionActivatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: AudioSessionActivatedEvent.name
        )
      }
    }

    OnStartObserving(AudioSessionDeactivatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: AudioSessionDeactivatedEvent.name
        )
      }
    }
    OnStopObserving(AudioSessionDeactivatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: AudioSessionDeactivatedEvent.name
        )
      }
    }

    OnStartObserving(AudioRouteChangedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: AudioRouteChangedEvent.name
        )
      }
    }
    OnStopObserving(AudioRouteChangedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: AudioRouteChangedEvent.name
        )
      }
    }

    OnStartObserving(IncomingCallReportedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: IncomingCallReportedEvent.name
        )
      }
    }
    OnStopObserving(IncomingCallReportedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: IncomingCallReportedEvent.name
        )
      }
    }

    OnStartObserving(OutgoingCallStartedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: OutgoingCallStartedEvent.name
        )
      }
    }
    OnStopObserving(OutgoingCallStartedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: OutgoingCallStartedEvent.name
        )
      }
    }

    OnStartObserving(CallAnsweredEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallAnsweredEvent.name
        )
      }
    }
    OnStopObserving(CallAnsweredEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallAnsweredEvent.name
        )
      }
    }

    OnStartObserving(CallEndedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallEndedEvent.name
        )
      }
    }
    OnStopObserving(CallEndedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallEndedEvent.name
        )
      }
    }

    OnStartObserving(CallReportedEnded.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallReportedEnded.name
        )
      }
    }
    OnStopObserving(CallReportedEnded.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallReportedEnded.name
        )
      }
    }

    OnStartObserving(SetMutedActionEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: SetMutedActionEvent.name
        )
      }
    }
    OnStopObserving(SetMutedActionEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: SetMutedActionEvent.name
        )
      }
    }

    OnStartObserving(VideoChangedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: VideoChangedEvent.name
        )
      }
    }
    OnStopObserving(VideoChangedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: VideoChangedEvent.name
        )
      }
    }

    OnStartObserving(SetHeldActionEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: SetHeldActionEvent.name
        )
      }
    }
    OnStopObserving(SetHeldActionEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: SetHeldActionEvent.name
        )
      }
    }

    OnStartObserving(DTMFEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: DTMFEvent.name
        )
      }
    }
    OnStopObserving(DTMFEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(eventName: DTMFEvent.name)
      }
    }

    OnStartObserving(CallIntentReceivedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: CallIntentReceivedEvent.name
        )
      }
    }
    OnStopObserving(CallIntentReceivedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: CallIntentReceivedEvent.name
        )
      }
    }

    OnStartObserving(VoIPPushTokenUpdatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.startObserving(
          eventName: VoIPPushTokenUpdatedEvent.name
        )
      }
    }
    OnStopObserving(VoIPPushTokenUpdatedEvent.name) {
      Task { @MainActor in
        CallEventEmitter.shared.stopObserving(
          eventName: VoIPPushTokenUpdatedEvent.name
        )
      }
    }

    // MARK: - Call Session

    AsyncFunction("getActiveCallSession") { () -> [String: Any]? in
      guard let session = await CallManager.shared.store.firstSession
      else {
        return nil
      }
      return session.toDictionary()
    }

    // MARK: - Audio Session

    Function("getAudioSessionState") { () -> [String: Any] in
      return AudioManager.shared.getAudioSessionState()
    }

    Function("setRTCAudioSessionConfiguration") { (hasVideo: Bool) in
      AudioManager.shared.setRTCAudioSessionConfiguration(hasVideo: hasVideo)
    }

    Function("prepareAudioSessionForCall") { (hasVideo: Bool) in
      AudioManager.shared.prepareAudioSessionForCall(hasVideo: hasVideo)
    }

    Function("restoreAudioSession") {
      AudioManager.shared.restoreAudioSession()
    }

    Function("setAudioSessionPortOverride") { (enabled: Bool) in
      AudioManager.shared.setAudioSessionPortOverride(enabled)
    }

    // MARK: - Capture Session

    Function("getCaptureSessionState") { () -> [String: Any] in
      return CaptureSessionManager.shared.getCaptureSessionState()
    }

    // MARK: - Start Outgoing Call

    AsyncFunction("startOutgoingCall") {
      (recipient: CallParticipantRecord, options: CallOptionsRecord)
        -> String in
      let id = try await CallManager.shared.startOutgoingCall(
        recipient: recipient.toModel(),
        options: options.toModel()
      )
      return id.uuidString
    }

    // MARK: - Report Incoming Call

    AsyncFunction("reportIncomingCall") {
      (event: IncomingCallEventRecord) in
      let incomingCallEvent = try event.toModel()
      try await CallManager.shared.reportIncomingCall(
        event: incomingCallEvent
      )
    }

    // MARK: - Answer Call

    AsyncFunction("answerCall") { (id: String) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      try await CallManager.shared.answerCall(for: uuid)
    }

    AsyncFunction("fulfillIncomingCallAnswered") { (requestId: String) in
      guard let requestUUID = UUID(uuidString: requestId) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid request ID: \(requestId)"
        )
      }

      await CallManager.shared.fulfillIncomingCall(requestId: requestUUID)
    }

    AsyncFunction("failIncomingCallConnected") { (requestId: String) in
      guard let requestUUID = UUID(uuidString: requestId) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid request ID: \(requestId)"
        )
      }

      await CallManager.shared.failIncomingCallConnected(requestId: requestUUID)
    }

    AsyncFunction("reportOutgoingCallConnected") { (id: String) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      await CallManager.shared.reportOutgoingCallConnected(for: uuid)
    }

    // MARK: - End Call

    AsyncFunction("endCall") { (id: String) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      try await CallManager.shared.endCall(for: uuid)
    }

    AsyncFunction("reportCallEnded") { (id: String, reason: String) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      let endedReason: CXCallEndedReason
      switch reason {
      case "failed": endedReason = .failed
      case "remoteEnded": endedReason = .remoteEnded
      case "unanswered": endedReason = .unanswered
      case "answeredElsewhere": endedReason = .answeredElsewhere
      case "declinedElsewhere": endedReason = .declinedElsewhere
      default: endedReason = .failed
      }

      await CallManager.shared.reportCallEnded(
        for: uuid,
        reason: endedReason
      )
    }

    // MARK: - Mute Support

    AsyncFunction("setMuted") { (id: String, muted: Bool) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      try await CallManager.shared.setMuted(for: uuid, muted: muted)
    }

    // MARK: - Video Support

    AsyncFunction("reportVideo") { (id: String, enabled: Bool) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      await CallManager.shared.reportVideo(for: uuid, enabled: enabled)
    }

    // MARK: - Hold Support

    AsyncFunction("setHeld") { (id: String, onHold: Bool) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      try await CallManager.shared.setHeld(for: uuid, onHold: onHold)
    }

    // MARK: - DTMF Support

    AsyncFunction("playDTMF") { (id: String, digits: String) in
      guard let uuid = UUID(uuidString: id) else {
        throw Exception(
          name: "InvalidUUID",
          description: "Invalid call ID: \(id)"
        )
      }

      try await CallManager.shared.playDTMF(for: uuid, digits: digits)
    }

    // MARK: - VoIP Push

    Function("registerVoIPPush") {
      Task { @MainActor in
        VoIPPushManager.shared.register()
      }
    }

    Function("getVoIPPushToken") { () -> [String: Any?] in
      return [
        "token": VoIPPushManager.shared.token as Any?,
        "type": "APNS_VOIP",
      ]
    }

  }
}
